### DISCUSSION TITLE
regression in Group by distinct count

### DISCUSSION THREAD

**Discussion by djouallah** [duckdb/duckdb#5165] [General]
this query is becoming slower, using sf=1
![image](https://user-images.githubusercontent.com/12554469/199365096-37dd69be-81be-487e-90f1-baa00ff9660e.png)

in my laptop with sf=10, I am getting out of memory 
see reproducible notebook
https://colab.research.google.com/drive/1ybdfVJdZq5HVRwm7J9OjbLXxYwyaqOdc#scrollTo=xAGX9tCqyGeU

**Comment by hannes** [comment]
Maybe @Tishj can have a look here?

**Reply by Tishj** [reply]
It's probably related to the same thing we're still looking to improve on the benchmark performance - parallelizing the scan over the distinct ht's 

I'll have to do some profiling to be sure, but my guess is that this has a lot of distinct values

**Reply by Mytherin** [reply]
For reference, here are the timings on my laptop:

```
master:
8T -> 1.02s
1T -> 1.57s

0.5.1: 
1T -> 1.17s
8T -> 1.17s
```

After the parallel scans over the distinct HTs are implemented this should be significantly faster in the multi-thread case.

**Reply by djouallah** [reply]
now getting crash , when using those parquet files with this Query
https://drive.google.com/drive/folders/1mZC3NuPBZC4mjP3_kH18c9fLrv8ME7RU

```
SELECT
    l_comment ,   count (distinct L_orderkey),  
FROM     lineitem
group by 1
order by 2 desc 
limit 10 ;
```
![image](https://user-images.githubusercontent.com/12554469/199730353-9aea6dec-3702-4dc7-86e7-9082bcb7c530.png)

**Reply by Mytherin** [reply]
Are you getting a crash or running out of memory?

**Reply by djouallah** [reply]
not sure maybe memory, I don't know

**Reply by Mytherin** [reply]
What do the logs say?

**Reply by djouallah** [reply]
error 23:40:18.283: Disposing session as kernel process died ExitCode: 3221225477, Reason: c:\Users\mimoune.djouallah\AppData\Local\Programs\Python\Python310\lib\site-packages\traitlets\traitlets.py:2392: FutureWarning: Supporting extra quotes around strings is deprecated in traitlets 5.0. You can use 'hmac-sha256' instead of '"hmac-sha256"' if you require traitlets >=5.
  warn(
c:\Users\mimoune.djouallah\AppData\Local\Programs\Python\Python310\lib\site-packages\traitlets\traitlets.py:2346: FutureWarning: Supporting extra quotes around Bytes is deprecated in traitlets 5.0. Use '458e0f0e-c520-490c-a4ae-ff33e53b532c' instead of 'b"458e0f0e-c520-490c-a4ae-ff33e53b532c"'.
  warn(

info 23:40:18.316: Dispose Kernel process 15080.
error 23:40:18.330: Raw kernel process exited code: 3221225477
error 23:40:18.428: Error in waiting for cell to complete [Error: Canceled future for execute_request message before replies were done
	at t.KernelShellFutureHandler.dispose (c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:2:32353)
	at c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:2:51405
	at Map.forEach (<anonymous>)
	at y._clearKernelState (c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:2:51390)
	at y.dispose (c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:2:44872)
	at c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:2:2218254
	at t.swallowExceptions (c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:7:130943)
	at p.dispose (c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:2:2218232)
	at t.RawSession.dispose (c:\Users\mimoune.djouallah\.vscode\extensions\ms-toolsai.jupyter-2022.9.1202862440\out\extension.node.js:2:2223340)
	at process.processTicksAndRejections (node:internal/process/task_queues:96:5)]
warn 23:40:18.459: Cell completed with errors {
  message: 'Canceled future for execute_request message before replies were done'
}
info 23:40:18.509: Cancel all remaining cells true || Error || undefined
warn 23:40:18.578: 2022-11-03 23:40:18,553 UTC - WARNING - Unknown command: DISPOSE_INTERRUPT_HANDLE:1:1352

**Reply by djouallah** [reply]
and it does spill to disk
![image](https://user-images.githubusercontent.com/12554469/199738162-97b951ad-7bbb-4ddc-b4a8-b5c50e4ecace.png)

**Reply by Mytherin** [reply]
The query seems to run fine for me locally, both running straight from the Parquet file and running from an in-memory database. If I decrease my memory limit to 4GB I get an out-of-memory error as expected:

```sql
Error: Out of Memory Error: could not allocate block of 262144 bytes (4000264448/4000000000 used) 
```

The OOM killer killing the kernel instead of providing a proper out-of-memory error might be related to the jemalloc issue. In general, it is expected that aggregate hashtables are not spilled to disk yet, so it is not surprising that this query does not finish in a low-memory situation.

**Reply by djouallah** [reply]
it works fine in 5.1 
![image](https://user-images.githubusercontent.com/12554469/199742379-655dd198-0530-4686-b65a-685d088a3bb0.png)

**Reply by Mytherin** [reply]
Do you still run out of memory if you disable multi-threading (`SET threads=1`)? Running with multiple threads generally uses more memory - and if the query is already very close to the memory limit of your machine the new multi-threading capabilities might well push it over the edge.

**Reply by djouallah** [reply]
it does work now in master with thread=1
![image](https://user-images.githubusercontent.com/12554469/199744392-e7da0059-1e97-4b94-be27-5371319d69ac.png)

**Reply by Mytherin** [reply]
Alright, so then I would say this is not a regression, it is just caused by the new multi-threading mode using additional memory.

**Reply by djouallah** [reply]
yes, but it is a still a problem, it is using a lot of memory !!!

**Reply by Mytherin** [reply]
Well, you are grouping by an almost distinct string column of a 10GB database on a machine with 7GB of RAM. Out-of-core hash aggregates are on the roadmap, but not there yet. Until they are implemented you will run into out-of-memory scenarios when your `GROUP BY` sets do not fit in memory. In my experience almost all other systems have similar restrictions.

**Reply by djouallah** [reply]
just for the record my laptop has 16 GB of RAM, and Vertipaq render it in 29 second.

**Reply by Mytherin** [reply]
The query for me requires 10GB of RAM in single-threaded mode and 14GB of RAM in multi-threaded mode. In v0.5.1 it always required 10GB of RAM (since there was only a single-threaded mode). I'm guessing the standard memory limit of your laptop is barely exceeded by the query when multi-threading. We set the limit relatively conservative to 80% of system RAM. You could try increasing the limit (e.g. `SET memory_limit='15GB'`), or just disabling the memory limit entirely (`SET memory_limit=-1`). It should complete then.

**Reply by djouallah** [reply]
make sense, yes, it works when I removed the memory limit, but, it is definitely slower than a single thread ?
![image](https://user-images.githubusercontent.com/12554469/199853945-ed6089fd-1c1f-4d77-8632-4033a700e962.png)

**Reply by Tishj** [reply]
That doesn't look right, I'll have to see what's happening there
This is using a commit after the last change I made to parallelize the distinct aggregations?

**Reply by djouallah** [reply]
using 0.5.2.dev1575

**Reply by djouallah** [reply]
@Tishj the weird behavior is only when I query using this parquet file
https://drive.google.com/drive/folders/1mZC3NuPBZC4mjP3_kH18c9fLrv8ME7RU

**Reply by djouallah** [reply]
today update, no memory error, although it is slow
![image](https://user-images.githubusercontent.com/12554469/200238668-66fc28ea-014f-4eff-bc27-5bfb6b5ed863.png)
